src/host/_output.cpp

```diff
@@ -53,6 +50,208 @@ void WriteToScreen(SCREEN_INFORMATION& screenInfo, const Viewport& region)
+    // However, OutputCellIterator is terrifyingly unsafe code and so we don't do that.
+    //
+    // Constructing an OutputCellIterator with a `wchar_t` takes the `wchar_t` by reference, so that it can reference
+    // it in a `wstring_view` forever. That's of course really bad because passing a `const uint16_t&` to a
+    // `const wchar_t&` argument implicitly converts the types. To do so, the implicit conversion allocates a
+    // `wchar_t` value on the stack. The lifetime of that copy DOES NOT get extended beyond the constructor call.
+    // The result is that OutputCellIterator would read random data from the stack.
+    //
+    // Don't ever assume the lifetime of implicitly convertible types given by reference.
+    // Ironically that's a bug that cannot happen with C pointers. To no ones surprise, C keeps on winning.
+    auto attrs = static_cast<const uint16_t*>(data);
+    auto chars = static_cast<const wchar_t*>(data);
+
+    if (!bufferSize.IsInBounds(startingCoordinate))
+    {
+        return {};
+    }
+
+    if (const auto io = gci.GetVtIoForBuffer(&screenInfo))
+    {
+        const auto corkLock = io->Cork();
+
+        const auto h = bufferSize.Height();
+        const auto w = bufferSize.Width();
+        auto y = startingCoordinate.y;
+        til::CoordType end = 0;
+        auto remaining = lengthToWrite;
+
+        til::small_vector<CHAR_INFO, 1024> infos;
+        infos.resize(gsl::narrow_cast<size_t>(w) + 1);
+
+        Viewport unused;
+
+        while (y < h && remaining > 0)
+        {
+            const auto beg = y == startingCoordinate.y ? startingCoordinate.x : 0;
+            auto len = std::min(remaining, gsl::narrow_cast<size_t>(w - beg));
+            end = beg + gsl::narrow_cast<til::CoordType>(len);
+
+            auto viewport = Viewport::FromInclusive({ beg, y, end - 1, y });
+            THROW_IF_FAILED(ReadConsoleOutputWImplHelper(screenInfo, infos, viewport, unused));
+
+            switch (mode)
+            {
+            case FillConsoleMode::WriteAttribute:
+                for (size_t i = 0; i < len; ++i)
+                {
+                    infos[i].Attributes = *attrs++;
+                }
+                break;
+            case FillConsoleMode::WriteCharacter:
+                for (size_t i = 0; i < len;)
+                {
+                    const auto ch = *chars++;
+
+                    auto& lead = infos[i++];
+                    lead.Char.UnicodeChar = ch;
+                    lead.Attributes = lead.Attributes & ~(COMMON_LVB_LEADING_BYTE | COMMON_LVB_TRAILING_BYTE);
+
+                    if (IsGlyphFullWidth(ch))
+                    {
+                        lead.Attributes |= COMMON_LVB_LEADING_BYTE;
+
+                        auto& trail = infos[i++];
+                        trail.Char.UnicodeChar = ch;
+                        trail.Attributes = trail.Attributes & ~(COMMON_LVB_LEADING_BYTE | COMMON_LVB_TRAILING_BYTE) | COMMON_LVB_LEADING_BYTE;
+                    }

```

Comment:
Yep, that's a bug. I'll add this to the unit tests. I've also changed the `WriteCharacter` code to be identical to `FillCharacter` so that it's easier to read.